### PR TITLE
Made !uqfinal dynamic by semester, added !dog (like !cat) and !cookbook (link)

### PR DESCRIPTION
**Pull request by nickl93**
All the feedback plz. ❤️

### CODE REVIEW

**Comment by mitchmcdee** [comment]
it's no biggie, but in future you should probably split this up into 3 PRs so that we can seperate the features and give feedback/approval accordingly :)

File: scripts/dog.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Description
+//   Like !cat, but for dog people (and with less trauma for 2310 students
```
Comment (mitchmcdee) [inline scripts/dog.js]: forgot the ')' at the end :^)

Comment (mitchmcdee) [reply scripts/dog.js]: also, could you provide a screenshot of the dog printout? seeing as though its a visual script, it'd be nice to see what it looks like for whatever reason :)

File: scripts/uqfinal.js [hunk 2]
```diff
@@ -8,58 +8,90 @@
...
+          res.send("Failure fetching current semester (HTTP Error: " + resp.statusCode + ")");
+          return;
+        }
+
+        var semesters = JSON.parse(body).data.semesters;
+        for (var i = 0; i < semesters.length; i++) {
+          if (semesters[i].isCurrent) {
+            semester = semesters[i].uqId;
+            break;
+          } else {
```
Comment (mitchmcdee) [inline scripts/uqfinal.js]: dangling else statement :)

File: scripts/uqfinal.js [hunk 3]
```diff
@@ -8,58 +8,90 @@
...
-						var result = Math.ceil(needed/course.assessment[num_of_assessment - 1].weight * 100);
-						res.send("You need to achieve at least " + result + "% on the final exam.\n_Disclaimer: this does not take hurdles into account_\n_Powered by http://uqfinal.com_");
-					}
-				});
-	});
-}
+  robot.respond(/!?uqfinal ([a-z]{4}[0-9]{4}[a-z]?) (.+)+/i, function (res) {
+    //res.match[1] is the course code
+    //res.match[2] is the space-seperated scores
+    var semester;
```
Comment (mitchmcdee) [inline scripts/uqfinal.js]: does this need to be hiding out here? I (personally) think it would give context to the `if (!semester)` code below to have it there

File: scripts/dog.js (hunk 1 above)
Comment (nickl93) [reply scripts/dog.js]: screen shot 2017-11-01 at 8 04 54 pm

File: scripts/uqfinal.js [hunk 4]
```diff
@@ -8,58 +8,98 @@
...
+        }
+
+        // Handle failed HTTP request (check API endpoints at uqfinal.com if consistently failing)
+        if (resp.statusCode != 200) {
+          res.send("Failure fetching current semester (HTTP Error: " + resp.statusCode + ")");
+          return;
+        }
+
+        // Get the current semester uqId (as per UQ Rota)
+        var semesters = JSON.parse(body).data.semesters;
```
Comment (mitchmcdee) [inline scripts/uqfinal.js]: should probably double check the API/gricey, but on quick inspection it looks like its returning a sorted list of semesters-- so the current semester will always be the last-- so why don't we just do `JSON.parse(body).data.semesters.pop()` or something?

File: package.json [hunk 5]
```diff
@@ -18,6 +18,7 @@
     "ical": "^0.5.0",
     "ical.js": "^1.2.2",
     "mathjs": "^3.2.1",
+    "purescript": "^0.11.6",
```
Comment (mitchmcdee) [inline package.json]: why was this added?

Comment (nickl93) [reply package.json]: My bad, that was me trying to get the purescript issues to go away (until trm provided a fix). Accidentally deleting package.json was me (poorly) trying to stop this from being included.

Comment (mitchmcdee) [reply package.json]: bahaha no wozzas

File: scripts/uqfinal.js [hunk 6]
```diff
@@ -8,58 +8,92 @@
...
+        }
+
+        // Handle failed HTTP request (check API endpoints at uqfinal.com if consistently failing)
+        if (resp.statusCode != 200) {
+          res.send("Failure fetching current semester (HTTP Error: " + resp.statusCode + ")");
+          return;
+        }
+
+        // Get the current semester uqId (as per UQ Rota)
+        var semester = JSON.parse(body).data.semesters.pop().uqId;
```
Comment (mitchmcdee) [inline scripts/uqfinal.js]: this will error if `pop()` returns undefined for whatever reason. I'd just put the `uqId` in the `robot.http()` request below :)

File: scripts/cookbook.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Description
+//   Returns a link to the UQCS student-compiled cookbook (full of tasty treats).
+// Commands
+//   `!cookbook` - returns the URL of the UQCS student-compiled cookbook (pdf).
+//
+
+module.exports = function (robot) {
+  robot.respond(/!?cookbook/i, function (res) {
+    res.send("https://jenkins.uqcs.org.au/job/cookbook/lastSuccessfulBuild/artifact/main.pdf");
```
Comment (TRManderson) [inline scripts/cookbook.js]: Can you make it link to https://github.com/UQComputingSociety/cookbook instead? there's a link to the compiled PDF there but it has all the extra info too, so it's probs a better option.
